boards/arm/thingy91_nrf9160/nrf52840_reset.c

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/gpio.h>
+#include <drivers/uart.h>
+#include <device.h>
+
+#define RESET_NODE DT_NODELABEL(nrf52840_reset)
+
+#if DT_NODE_HAS_STATUS(RESET_NODE, okay)
+
+#define RESET_GPIO_CTRL  DT_GPIO_CTLR(RESET_NODE, gpios)
+#define RESET_GPIO_PIN   DT_GPIO_PIN(RESET_NODE, gpios)
+#define RESET_GPIO_FLAGS DT_GPIO_FLAGS(RESET_NODE, gpios)
+
+int bt_hci_transport_setup(const struct device *h4)
+{
+	int err;
+	char c;
+	const struct device *port = DEVICE_DT_GET(RESET_GPIO_CTRL);
+
+	if (!device_is_ready(port)) {
+		return -EIO;
+	}
+
+	/* Configure pin as output and initialize it to inactive state. */
+	err = gpio_pin_configure(port, RESET_GPIO_PIN,
+				 RESET_GPIO_FLAGS | GPIO_OUTPUT_INACTIVE);
+	if (err) {
+		return err;
+	}
+
+	/* Reset the nRF52840 and let it wait until the pin is inactive again
+	 * before running to main to ensure that it won't send any data until
+	 * the H4 device is setup and ready to receive.
+	 */
+	err = gpio_pin_set(port, RESET_GPIO_PIN, 1);
+	if (err) {
+		return err;
+	}
+
+	/* Wait for the nRF52840 peripheral to stop sending data.
+	 *
+	 * It is critical (!) to wait here, so that all bytes
+	 * on the lines are received and drained correctly.
+	 */
+	k_sleep(K_MSEC(10));
```

Comment:
Is it a boot banner sending on UART after nrf52 reset ?
If so we can disable it.

Reply:
Not sure if that's the issue. Someone else can likely comment on this, but my guess is that the hci_lpuart application tries to communicate with the nRF9160 before synchronizing. The data sent during that period must be purged after the nRF52840 is reset, so that the application starts with no garbage on the lines.

Reply:
> When the nRF9160 is reset, we need to reset the nRF52840 chip as well to synchronize HCI. We reset the nRF52 using the pin above, and we want to wait for whatever was being sent by the nRF52 to appear on the UART line and drain it, before we let the nRF52 run again to main and re-initialize HCI.

I have explained this already. It's not the boot banner, the nRF52 chip is sending HCI data and it is reset when the nRF91 is reset so that HCI can be re-initialized cleanly. After the nRF52 chip is reset we need to wait for whatever was being sent on the UART to appear on the other side, and "drain" it, otherwise we will fail HCI initialization. If you remove the sleep you will fail HCI initialization when you reset the nRF91.